### ISSUE TITLE
acfql actor loss issue

### ISSUE THREAD

**Issue by SitheralS** [ColinQiyangLi/qc#9]
Hi, thanks for providing such a structured and clear repo. 
When I tried to replicate your results using pytorch, I find your actor loss is:

    actor_loss = bc_flow_loss + self.config['alpha'] * distill_loss + q_loss

where:

    target_flow_actions = self.compute_flow_actions(batch['observations'], noises=noises)
    actor_actions = self.network.select('actor_onestep_flow')(batch['observations'], noises, params=grad_params)
    distill_loss = jnp.mean((actor_actions - target_flow_actions) ** 2)

However, when doing the back-propagation, it seems that you haven't stopped the gradient of computing target_flow_actions, which means that you are not just training the actor_one_step_flow, but simultaneously training actor_one_step_flow and actor_BC with two losses (bc_flow_loss and q_loss), while the distill_loss simply serves for aligning the two actors. Isn't it not consistent with your paper's idea?

I have tried replicating your results using pytorch, for the task of robomimic-square (but here I used square-mh-image). It seems that I could successfully reach about 80% success rate using your code's gradient computing method (gradient on for the actor_BC), but the model would fail if I switch to your paper's gradient computing method (gradient off for the actor_BC). Does this mean that I have some misunderstandings about your paper's method and that your realization in this repo is exactly what you proposed in your paper?

**Comment by ColinQiyangLi** [comment]
Hi, thanks for raising the question.

In the `self.compute_flow_actions` function, the velocity is computed without passing in `params`, so there is no gradient flowing through`target_flow_actions`:

https://github.com/ColinQiyangLi/qc/blob/7da987d93a254b9bdf5d949bbfbcb5200e109e84/agents/acfql.py#L220

Hope this clarifies the implementation!

Re: square-mh-image -- that is interesting.. Thanks for sharing your results! I guess if you enable the gradient through the `target_flow_actions`, the BC policy can get pulled away from the behavior distribution a bit, effectively decreasing the strength of the behavior constraint. Have you tried slightly different alpha coefficients?

**Comment by SitheralS** [comment]
Hi, thanks for the reply!

It turns out that it was because of a mistake of my code that the behavior cloning policy didn't function as expected. After I corrected the problems, I could then replicate the results. 

About the alpha, since I am using the sparse 0/1 reward instead of your original -1/0 reward, I think alpha=1 would always be good.

Thanks again for the clarification!
